vars/artifactSetVersion.groovy

```diff
@@ -0,0 +1,138 @@
+            'versioningTemplate'
+        ]
+
+        Map configuration = ConfigurationMerger.mergeWithPipelineData(parameters, parameterKeys, pipelineDataMap, stepConfiguration, stepConfigurationKeys, stepDefaults)
+
+        //buildTool is mandatory parameter
+        def utils = new Utils()
+        def buildTool = utils.getMandatoryParameter(configuration, 'buildTool')
+
+        if (!configuration.filePath)
+            configuration.filePath = configuration[buildTool].filePath //use default configuration
+
+        def newVersion
+        def artifactVersioning = ArtifactVersioning.getArtifactVersioning(buildTool, this, configuration)
+
+        if(configuration.artifactType == 'appContainer' && configuration.dockerVersionSource == 'appVersion'){
+            if (script.commonPipelineEnvironment.getArtifactVersion())
+                //replace + sign if available since + is not allowed in a Docker tag
+                newVersion = script.commonPipelineEnvironment.getArtifactVersion().replace('+', '_')
+            else
+                error ("[${stepName}] No artifact version available for 'dockerVersionSource: appVersion' -> executeBuild needs to run for the application artifact first to set the artifactVersion for the application artifact.'")
+        } else {
+            def currentVersion = artifactVersioning.getVersion()
+
+            def timestamp = configuration.timestamp ? configuration.timestamp : getTimestamp(configuration.timestampTemplate)
+
+            def versioningTemplate = configuration.versioningTemplate ? configuration.versioningTemplate : configuration[configuration.buildTool].versioningTemplate
+            //defined in default configuration
+            def binding = [version: currentVersion, timestamp: timestamp, commitId: configuration.gitCommitId]
+            def templatingEngine = new SimpleTemplateEngine()
+            def template = templatingEngine.createTemplate(versioningTemplate).make(binding)
+            newVersion = template.toString()
+        }
+
+        artifactVersioning.setVersion(newVersion)
+
+        sh 'git add .'
+
+        def gitCommitId
+
+        sshagent([configuration.gitCredentialsId]) {
+            if (configuration.gitUserName  && configuration.gitUserEMail) {
+                sh "git config user.name \"${configuration.gitUserName}\""
+                sh "git config user.email \"${configuration.gitUserEMail}\""
+            }
+            try {
+                sh "git commit -m 'update version ${newVersion}'"
+            } catch (e) {
+                echo "[${stepName}]git commit failed: ${e}"
+            }
+            sh "git remote set-url origin ${configuration.gitSshUrl}"
+            sh "git tag ${configuration.tagPrefix}${newVersion}"
+            sh "git push origin ${configuration.tagPrefix}${newVersion}"
+
+            gitCommitId = gitUtils.getGitCommitId()
+
+        }
+
+        if(buildTool == 'docker' && configuration.artifactType == 'appContainer') {
```

Comment:
I recommend to put everything which is specific to a build toolset in our build toolset specific versioning classes. Otherwise the if () {} else if() {} else if {} gets longer and longer.

Reply:
As indicated above this is due to the specific Docker build pattern and other cases are not foreseen.